### PR TITLE
agent: add GetGuestDetails gRPC function

### LINKED ISSUE #354: agent: get content of a file in sandbox
# Description of problem

In https://github.com/kata-containers/runtime/pull/624, we have such problems:
1. Hot add memory should be aligned to memory section size, but the size of a memory section is architecture dependent. For example, power uses 16MB, ia64 uses 1GB, x86_64 uses 128MB, ppc64le uses 256MB.
2. Each memory block is described under /sys/devices/system/memory as /sys/devices/system/memory/memoryXXX (XXX is the memory block id.), which is the unit of memory online/offline. In sparse memory model, memory block size is a multiple of memory section size, [reference](http://www.infradead.org/~mchehab/kernel_docs/unsorted/memory-hotplug.html#unit-of-memory-online-offline-operation). We can online a memory block only if we hotplug a memory-block-sized dimm, see `pages_correctly_reserved` in [reference](https://elixir.bootlin.com/linux/v4.14.49/source/drivers/base/memory.c#L230).

So we should get the size of guestos memory block by reading /sys/devices/system/memoryblock_size_bytes in sandbox, and make hot add memory to be aligned to memory block size of sandbox. 

I think it's needed to add a ReadSandboxFile rpc to read file above in a common way.

### PR DESCRIPTION
**Pull request by linzichang**
Fixes #354

### CODE REVIEW

**Comment by jodh-intel** [comment]
Let's get input from others on whether this approach is a good one first.

/cc @sboeuf, @bergwolf, @jon.

**Comment by sboeuf** [comment]
@linzichang Thanks for the PR, but it looks a bit hacky to me to use `Check()` to retrieve the amount of memory from the guest. I think introducing a new function dedicated to this would be more appropriate.

I like the approach since we would rely on actual data from the guest OS, instead of having some hardcoded value from each hypervisor implementation, depending on the architecture.

That being said, we need a fallback, by proposing some default values, in case the grpc function is not supported by the agent, or return a non valid value for some reason.

**Comment by linzichang** [comment]
@sboeuf So you think we should:
1. Add a new grpc function like `getMemBlockSize`.
2. In kata-runtime side, give a default values if agent don't support or fail for `getMemBlockSize`.

**Comment by jodh-intel** [comment]
@linzichang - I think that's what @sboeuf is saying, yes. I have to admit, I'm still unclear about the gRPC best practice we're encountering here - should we be adding as many discrete functions as we need? Or should we be trying to provide more general functions that could return the values we need.

In this case, the latter approach might be to create a new call like:

```
message GuestDetailsResponse {
    uint64 block_size_bytes = 1;

    // XXX: add more fields in future when we discover we actually need more guest information.
}

rpc GetGuestDetails() returns (GuestDetailsResponse);
```

Of course if each gRCP call took a significant amount of time, we'd be more motivated to go for the general approach. Although we now have tracing in the runtime and although that includes gRCP client flows, it's not that easy to determine how much time is attributable directly to the protocol handling itself. As noted in https://github.com/kata-containers/kata-containers/issues/27#issuecomment-414677578, the first gRPC call (`Check` health check request) is extremely *slow* as I think it's waiting for the gRPC server in the agent to be available (aka waiting for the VM to boot)).

However, I added a second call to that essentially "empty" / NOP call  in `virtcontainers/kata_agent.go` and that takes between `0.7ms` and `1ms` on an i5 NUC which gives a very crude estimation of the end-to-end gRPC call time.

/cc @grahamwhaley.

**Comment by linzichang** [comment]
@jodh-intel Like we talk in #357 , it's very wonderful for some experts to give a gRPC design guide for our project. In this single case, we can just add a `getGuestInfo` to get this done.

**Comment by bergwolf** [comment]
@linzichang Could you do this in a separate gRPC call? The check gRPC is too frequent for such a one time attribute query. What about something like,
```
diff --git a/protocols/grpc/agent.proto b/protocols/grpc/agent.proto
index 4c09d9c..40e10e4 100644
--- a/protocols/grpc/agent.proto
+++ b/protocols/grpc/agent.proto
@@ -54,6 +54,7 @@ service AgentService {
        rpc DestroySandbox(DestroySandboxRequest) returns (google.protobuf.Empty);
        rpc OnlineCPUMem(OnlineCPUMemRequest) returns (google.protobuf.Empty);
        rpc ReseedRandomDev(ReseedRandomDevRequest) returns (google.protobuf.Empty);
+       rpc QueryResource(QueryResourceRequest) returns (QueryResourceResponse);
 }

 message CreateContainerRequest {
@@ -350,6 +351,18 @@ message ReseedRandomDevRequest {
        bytes data = 2;
 }

+message QueryResourceRequest {
+       // MemBlockSize asks server to return the system memory block size that can be used
+       // for memory hotplug alignment. Typically the server returns what's in
+       // /sys/devices/system/memory/block_size_bytes.
+       bool mem_block_size = 1;
+}
+
+message QueryResourceResponse {
+       // MemBlockSizeBytes returns the system memory block size in bytes.
+       uint64 mem_block_size_bytes = 1;
+}
+
 // Storage represents both the rootfs of the container, and any volume that
 // could have been defined through the Mount list of the OCI specification.
 message Storage {
```
Then kata-runtime only needs to query when necessary. And the API is extensible for future needs if we want to query more resource attributes.

**Comment by linzichang** [comment]
@bergwolf Yes, after discuss in this conversation, we all decide to make new gRPC call. I will rework on this PR.

**Comment by jodh-intel** [comment]
Thanks @linzichang - sorry for the extra work as it was my suggestion to re-use (aka abuse) `Check()` :)

@bergwolf's - I like your suggested solution. btw, if you use the magic ` ```diff `, you get pretty coloured diff output :smile:

**Comment by bergwolf** [comment]
@linzichang Please fix CI as the mock server also needs to implement the new grpc API. Other than that it looks good to me. Thanks!

File: grpc_test.go [hunk 1]
```diff
@@ -620,3 +620,19 @@ func TestMultiWaitProcess(t *testing.T) {
...
+		sandbox: &sandbox{
+			containers: make(map[string]*container),
+		},
+	}
+
+	req := &pb.QueryResourceRequest{
+		MemBlockSize: true,
+	}
+
+	_, err := a.QueryResource(context.TODO(), req)
```
Comment (jodh-intel) [inline grpc_test.go]: How about also comparing the return value with `/sys/devices/system/memory/block_size_bytes`?

**Comment by jodh-intel** [comment]
Just one test suggestion.

lgtm

File: grpc_test.go (hunk 1 above)
Comment (linzichang) [reply grpc_test.go]: I think about this too but it's dispensable. I will update it.

**Comment by linzichang** [comment]
PTAL

**Comment by devimc** [comment]
LGTM

**Comment by jodh-intel** [comment]
Based on the discussion we just had in the Architecture Meeting, we probably want to change this to a more generic name as outlined in https://github.com/kata-containers/agent/pull/358#issuecomment-419047830.

/cc @gnawux, @bergwolf, @egernst, @sboeuf.

**Comment by sboeuf** [comment]
Yes I agree. @linzichang the PR looks good and is almost ready to be merged, but please change `QueryResource()` to `GuestDetails()` as it is more generic naming.

**Comment by bergwolf** [comment]
@sboeuf @jodh-intel I'd like to keep the leading `Query` action word in the name. It makes it clear that the RPC is a query action that does not change anything in the guest, and all the other gRPCs are named as actions.

**Comment by linzichang** [comment]
Yes, I think `QueryGuestDetails` is better.

**Comment by linzichang** [comment]
The latest kata-runtime build failed.

**Comment by bergwolf** [comment]
@linzichang yeah, see https://github.com/kata-containers/runtime/pull/710

**Comment by jodh-intel** [comment]
a `Query` prefix is fine with me, but `Get` is shorter ;)

**Comment by linzichang** [comment]
CI is OK.

**Pull request merged**